### PR TITLE
Use mimc-based hash function for nested verification key digest

### PR DESCRIPTION
**Pull request by dtebbs**
- Use the mimc-based hash for the nested verification key.
- Requires a modification fo the verification key object, and a `scalar` versio nof the vk hashing gadget
- Large number of bit variables in VK are no longer required, so `scalar` versions of these objects have been added.

### CODE REVIEW

File: libzecale/tests/circuits/verification_key_hash_gadget_test.cpp [hunk 1]
```diff
@@ -101,6 +229,15 @@ void verification_key_hash_gadget_test()
     ASSERT_EQ(nvk_hash_value, hash_value);
 }
 
+TEST(VerificationKeyHashTest, HashTest)
+{
+    using wpp = libff::bw6_761_pp;
+    verification_key_hash_test<
+        wpp,
+        libzecale::groth16_verifier_parameters<wpp>,
```
Comment (AntoineRondelet) [inline libzecale/tests/circuits/verification_key_hash_gadget_test.cpp:237]: What about renaming the `nverifierT` template param as `nverifierParamsT` to better align with the actual use?

File: libzecale/circuits/groth16_verifier/r1cs_gg_ppzksnark_verifier_gadget.hpp [hunk 2]
```diff
@@ -97,6 +87,40 @@ class r1cs_gg_ppzksnark_verification_key_variable
             &r1cs_vk);
 };
 
+/// A version of r1cs_gg_ppzksnark_verification_key_variable without variables
+/// for the bits. In the case where an arithmetic hash of the verification key
+/// is used, this type saves many unnecessary variables.
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply libzecale/circuits/groth16_verifier/r1cs_gg_ppzksnark_verifier_gadget.hpp:92]: reworded.  thanks.

File: client/zecale/cli/zecale_nested_verification_key_hash.py [hunk 3]
```diff
@@ -26,11 +26,14 @@ def nested_verification_key_hash(
...
+    # zk-snark still uses the full width hash, which is returned by the
+    # aggregation server as a primary input when a batch is created).
     vk_hash_evm = list(hex_to_uint256_list(vk_hash_hex))
     vk_hash = vk_hash_evm[-1]
-    for i in vk_hash_evm[:-1]:
-        assert i == 0, "expected higher orders words to be 0"
+
+    # TODO: pass the full-width vk_hash to to the application (requires the
+    # dispatch interface to be updated), or show that using the LO word
+    # is still collision resistant.
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply client/zecale/cli/zecale_nested_verification_key_hash.py:37]: The hash value is too large to fit in a single evm word, but for now we continue to use only the LO word, and we currently can't be sure that this is collision resistant. So we either need to switch the contract and interfaces to use the multi-word hash, or we need to show that using the LO word alone is collision resistant.  (I don't think the latter is feasible at the momet, so a later PR switches everything to use the full vk_hash value, at which point this comment is removed.)

File: libzecale/circuits/groth16_verifier/r1cs_gg_ppzksnark_verifier_gadget.tcc [hunk 4]
```diff
@@ -97,12 +97,113 @@ template<typename ppT> size_t r1cs_gg_ppzksnark_proof_variable<ppT>::size()
...
+    // now, to avoid replicating that order (which also depends on the G1 and
+    // G2 variable gadgets), we instantiate this gadget and extract the values
+    // of _all_vars.
+
+    const size_t num_primary_inputs = r1cs_vk.ABC_g1.rest.indices.size();
+
+    // Create a circuit with a single vk gadget, and
+    libsnark::protoboard<FieldT> pb;
+    r1cs_gg_ppzksnark_verification_key_scalar_variable<ppT> vk(
+        pb, num_primary_inputs, " translation_step_vk");
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply libzecale/circuits/groth16_verifier/r1cs_gg_ppzksnark_verifier_gadget.tcc]: updated the annotation

File: libzecale/tests/circuits/verification_key_hash_gadget_test.cpp (hunk 1 above)
Comment (dtebbs) [reply libzecale/tests/circuits/verification_key_hash_gadget_test.cpp:237]: `nverifierParamsT` would end up pretty long, so I'm not sure it's worth adding `Params`.  (`nverifierT` is currently used pretty consistently throughout the code.)

File: libzecale/tests/circuits/verification_key_hash_gadget_test.cpp [hunk 5]
```diff
@@ -15,8 +17,134 @@
...
 {
 
+template<typename wppT, typename nverifierT>
+static void verification_key_scalar_hash_test()
+{
+    using FieldT = libff::Fr<wppT>;
+    using npp = libzecale::other_curve<wppT>;
+    using nsnark = typename nverifierT::snark;
+
+    // Get 2 VKs for the dummy app, and determien the number of primary inputs.
```
[the remark this answers is not in the document]
Comment (dtebbs) [reply libzecale/tests/circuits/verification_key_hash_gadget_test.cpp]: fixed.  thanks.

File: client/zecale/cli/zecale_nested_verification_key_hash.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (AntoineRondelet) [reply client/zecale/cli/zecale_nested_verification_key_hash.py:37]: Ah I see, you did it in 2 times (the full check on the vk is done in #68 I guess). Got it.

File: libzecale/circuits/groth16_verifier/r1cs_gg_ppzksnark_verifier_gadget.tcc (hunk 4 above)
[the remark this answers is not in the document]
Comment (AntoineRondelet) [reply libzecale/circuits/groth16_verifier/r1cs_gg_ppzksnark_verifier_gadget.tcc]: Thanks.
BTW, I just realized that the comment above this block (i.e. `// Create a circuit with a single vk gadget, and`) is unfinished.
What did you mean here?

I guess you initially meant something along the lines:
```console
// Create a circuit with a single vk gadget, and extract all scalars composing the verification key
```

or did you mean something else?

File: libzecale/tests/circuits/verification_key_hash_gadget_test.cpp (hunk 1 above)
Comment (AntoineRondelet) [reply libzecale/tests/circuits/verification_key_hash_gadget_test.cpp:237]: > (nverifierT is currently used pretty consistently throughout the code.)

Indeed.
I was mostly arguing that what this is instantiated with is the parameters classes and not the "verifier gadgets" etc.
We can stick with that for now, that's not a big deal.

File: libzecale/circuits/groth16_verifier/r1cs_gg_ppzksnark_verifier_gadget.tcc (hunk 4 above)
[the remark this answers is not in the document]
Comment (dtebbs) [reply libzecale/circuits/groth16_verifier/r1cs_gg_ppzksnark_verifier_gadget.tcc]: Thanks.  I deleted the comment, as I realised it's actually contained in the comment above.

**Review by AntoineRondelet — APPROVED** [review]

**Pull request merged**